# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: media/audioipc/audioipc/src/lib.rs
# Commit: 0a7873a6b522
# Full Hash: 0a7873a6b522ce4aef9a32ae4a47f093ba33e0f4
# Author: Matthew Gregan <kinetik@flim.org>
# Date: 2020-01-26 09:48:16
# Regressor Bug: 1590249
# File Overlap Count: 3
# Description:
#   Bug 1590249 - Update audioipc to 86d49ddf.  r=chunmin
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D60615
# ==============================================================================

diff -r 36f6505b50c9 -r 0a7873a6b522 media/audioipc/audioipc/src/lib.rs
--- a/media/audioipc/audioipc/src/lib.rs	Sat Jan 25 21:30:14 2020 +0000
+++ b/media/audioipc/audioipc/src/lib.rs	Sat Jan 25 21:31:32 2020 +0000
@@ -41,6 +41,9 @@
 #[cfg(unix)]
 mod tokio_uds_stream;
 
+#[cfg(windows)]
+mod tokio_named_pipes;
+
 pub use crate::messages::{ClientMessage, ServerMessage};
 use std::env::temp_dir;
 use std::path::PathBuf;
@@ -53,6 +56,8 @@
 #[cfg(windows)]
 use std::os::windows::io::{FromRawHandle, IntoRawHandle};
 
+use std::cell::RefCell;
+
 // This must match the definition of
 // ipc::FileDescriptor::PlatformHandleType in Gecko.
 #[cfg(windows)]
@@ -61,11 +66,19 @@
 pub type PlatformHandleType = libc::c_int;
 
 // This stands in for RawFd/RawHandle.
-#[derive(Copy, Clone, Debug)]
-pub struct PlatformHandle(PlatformHandleType);
+#[derive(Clone, Debug)]
+pub struct PlatformHandle(RefCell<Inner>);
+
+#[derive(Clone, Debug)]
+struct Inner {
+    handle: PlatformHandleType,
+    owned: bool,
+}
 
 unsafe impl Send for PlatformHandle {}
 
+pub const INVALID_HANDLE_VALUE: PlatformHandleType = -1isize as PlatformHandleType;
+
 // Custom serialization to treat HANDLEs as i64.  This is not valid in
 // general, but after sending the HANDLE value to a remote process we
 // use it to create a valid HANDLE via DuplicateHandle.
@@ -76,7 +89,8 @@
     where
         S: serde::Serializer,
     {
-        serializer.serialize_i64(self.0 as i64)
+        let h = self.0.borrow();
+        serializer.serialize_i64(h.handle as i64)
     }
 }
 
@@ -92,7 +106,8 @@
     where
         E: serde::de::Error,
     {
-        Ok(PlatformHandle::new(value as PlatformHandleType))
+        let owned = cfg!(windows);
+        Ok(PlatformHandle::new(value as PlatformHandleType, owned))
     }
 }
 
@@ -112,49 +127,54 @@
 
 #[cfg(windows)]
 fn valid_handle(handle: PlatformHandleType) -> bool {
-    const INVALID_HANDLE_VALUE: PlatformHandleType = -1isize as PlatformHandleType;
     const NULL_HANDLE_VALUE: PlatformHandleType = 0isize as PlatformHandleType;
     handle != INVALID_HANDLE_VALUE && handle != NULL_HANDLE_VALUE
 }
 
 impl PlatformHandle {
-    pub fn new(raw: PlatformHandleType) -> PlatformHandle {
-        PlatformHandle(raw)
-    }
-
-    pub fn try_new(raw: PlatformHandleType) -> Option<PlatformHandle> {
-        if !valid_handle(raw) {
-            return None;
-        }
-        Some(PlatformHandle::new(raw))
+    pub fn new(raw: PlatformHandleType, owned: bool) -> PlatformHandle {
+        assert!(valid_handle(raw));
+        let inner = Inner {
+            handle: raw,
+            owned: owned,
+        };
+        PlatformHandle(RefCell::new(inner))
     }
 
     #[cfg(windows)]
     pub fn from<T: IntoRawHandle>(from: T) -> PlatformHandle {
-        PlatformHandle::new(from.into_raw_handle())
+        PlatformHandle::new(from.into_raw_handle(), true)
     }
 
     #[cfg(unix)]
     pub fn from<T: IntoRawFd>(from: T) -> PlatformHandle {
-        PlatformHandle::new(from.into_raw_fd())
+        PlatformHandle::new(from.into_raw_fd(), true)
     }
 
     #[cfg(windows)]
-    pub unsafe fn into_file(self) -> std::fs::File {
-        std::fs::File::from_raw_handle(self.0)
+    pub unsafe fn into_file(&self) -> std::fs::File {
+        std::fs::File::from_raw_handle(self.into_raw())
     }
 
     #[cfg(unix)]
-    pub unsafe fn into_file(self) -> std::fs::File {
-        std::fs::File::from_raw_fd(self.0)
+    pub unsafe fn into_file(&self) -> std::fs::File {
+        std::fs::File::from_raw_fd(self.into_raw())
     }
 
-    pub fn as_raw(self) -> PlatformHandleType {
-        self.0
+    pub unsafe fn into_raw(&self) -> PlatformHandleType {
+        let mut h = self.0.borrow_mut();
+        assert!(h.owned);
+        h.owned = false;
+        h.handle
     }
+}
 
-    pub unsafe fn close(self) {
-        close_platformhandle(self.0);
+impl Drop for PlatformHandle {
+    fn drop(&mut self) {
+        let inner = self.0.borrow();
+        if inner.owned {
+            unsafe { close_platformhandle(inner.handle) }
+        }
     }
 }
 